src/Builder/FrontendDialectTransformer.cpp

```diff
@@ -61,20 +61,17 @@ class FrontendGenImpl {
   mlir::Location UnknownLoc() { return mlir::UnknownLoc::get(&context_); }
 
   /*!
-   * Import an onnx input tensor type by determining and recording its type
-   * in a list of input tensor mlir types.
-   * @param input onnx input tensor ValueInfoProto.
-   * @param arg_types list of mlir types representing types of graph input.
+   * Import an onnx tensor type by determining and returning its type
+   * @param value_info onnx tensor ValueInfoProto.
    */
-  mlir::Type ImportInputTensorType(const onnx::ValueInfoProto &input) {
+  mlir::Type ImportTensorType(const onnx::ValueInfoProto &value_info) {
     std::vector<int64_t> dims;
-    auto shape_proto = input.type().tensor_type().shape();
-    auto input_tensor_legalized_name = legalize_name(input.name());
+    auto shape_proto = value_info.type().tensor_type().shape();
```

Comment:
I have similar concerns. I believe current output type is deduced via shape/type inference (as opposed to taken without checking from the onnx model specification). Even though the onnx model spec may record information about the output type, it seems like the type information we derived is going to be more accurate/reliable.

Reply:
I have updated the code so that the type is imported only if the shape is presented in protobuf.

> Even though the onnx model spec may record information about the output type, it seems like the type information we derived is going to be more accurate/reliable.

I disagree. Shape inference is not always reliable. For example, current onnx-mlir cannot infer shapes for mobilenetv2.

Reply:
Hi @daquexian, my concern is that according to the ONNX python API documentation:
https://github.com/onnx/onnx/blob/master/docs/PythonAPIOverview.md#creating-an-onnx-model-using-helper-functions

You can put in anything for output shape and make an onnx model out of it. That's why I'm dubious about its correctness. Besides, your patch forces shape inference to happen on the operation whose output is returned, which effectively will always override the output type imported anyways, so why bother importing the output type?

Reply:
@tjingrant In the often-encountered cases that shape inference is not applicable, the output shape from onnx model is still useful. I'm implementing onnx optimizer based on onnx-mlir, the shape inference is not necessary in my case (for example, shape inference may fail on some complicated models, but these models can still be optimized), but the output shape is always needed to translate the mlir back to protobuf and ensure no information is lost